MobileCenter/MobileCenter/MSCustomProperties.m

```diff
@@ -0,0 +1,82 @@
+#import "MSCustomProperties.h"
+#import "MSCustomPropertiesPrivate.h"
+#import "MSMobileCenterInternal.h"
+
+static NSString *const kKeyPattern = @"^[a-zA-Z][a-zA-Z0-9]*$";
+static NSString *const kValueNullErrorMessage = @"Custom property value cannot be null, did you mean to call clear?";
+
+@implementation MSCustomProperties
+
+@synthesize properties = _properties;
+
+- (instancetype)init {
+  if ((self = [super init])) {
+    _properties = [NSMutableDictionary new];
+  }
+  return self;
+}
+
+- (MSCustomProperties *)setString:(NSString *)value forKey:(NSString *)key {
+  if ([self isValidKey:key]) {
```

Comment:
I meant you can introduce private setObject and then use this for setString, setNumber and setDate.